src/contexts/field-focus.js

```diff
@@ -1,39 +1,107 @@
-import React, { createContext, useContext, useEffect, useState } from 'react';
+import React, { createContext, useContext, useEffect, useReducer } from 'react';
 import {LocationContext} from "./location";
+import {copyObject} from "../utils";
 
 const defaults = {
-  focusPath: null,
-  setFocusPath: () => {},
+  focus: {
+    path: [],
+    field: null,
+    on: null,
+    last: null,
+  },
+  changeFocus: () => {},
   availableFocusPaths: [],
 };
 
 const findUniqueFieldNames = (resourceObjects) => {
-  return [...resourceObjects
-    .map(resourceObject => resourceObject.getFieldnames())
-    .reduce((reduced, fields) => {
-      return fields.reduce((reduced, field) => reduced.add(field), reduced);
-    }, new Set([]))
-  ];
+  const reduceAvailableFieldPaths = crumb => (reduced, resourceObject) => {
+    const fields = resourceObject.getFieldnames();
+    return fields.reduce((reduced, field) => {
+      reduced = reduced.add(`${crumb}${field}`);
+      if (resourceObject.hasRelationship(field)) {
```

Comment:
This will fail if `this.raw.relationships` is undefined, this happened for me for `fieldName` name on `user--user`